File: DataStructures/C#/LinearDataStructures/LinkedList.cs
namespace LinearDataStructures
{
    using System;
    using System.Collections;
    using System.Collections.Generic;

    using Extensions;

    public class LinkedList<T> : ILinkedList<T>
    {
        private LinkedListNode<T> firstNode;
        private LinkedListNode<T> lastNode;

        public LinkedList()
        {
            this.Count = 0;
            this.firstNode = null;
            this.lastNode = null;
        }

        public LinkedList(IEnumerable<T> collection)
            : this()
        {
            collection.ForEach(this.AddLast);
        }

        public T First
        {
            get
            {
                if (this.firstNode == null)
                {
                    throw new ArgumentOutOfRangeException(paramName: "Count", message: "No elements in the list");
                }

                return this.firstNode.Value;
            }
        }

        public T Last
        {
            get
            {
                if (this.lastNode == null)
                {
                    throw new ArgumentOutOfRangeException(paramName: "Count", message: "No elements in the list");
                }

                return this.lastNode.Value;
            }
        }

        public int Count { get; private set; }

        public void AddFirst(T value)
        {
            var newNode = new LinkedListNode<T>(value);
            if (this.firstNode == null)
            {
                this.firstNode = newNode;
                this.lastNode = newNode;
            }
            else
            {
                newNode.Next = this.firstNode;
                this.firstNode.Previous = newNode;
                this.firstNode = newNode;
            }
            ++this.Count;
        }

        public void AddLast(T value)
        {
            var newNode = new LinkedListNode<T>(value);
            if (this.lastNode == null)
            {
                this.firstNode = newNode;
                this.lastNode = newNode;
            }
            else
            {
                newNode.Previous = this.lastNode;
                this.lastNode.Next = newNode;
                this.lastNode = newNode;
            }
            ++this.Count;
        }

        public void RemoveFirst()
        {
            if (this.firstNode == null)
            {
                throw new ArgumentNullException(paramName: "No elements to remove");
            }

            this.firstNode = this.firstNode.Next;
            if (this.firstNode != null)
            {
                this.firstNode.Previous = null;
            }

            --this.Count;
        }

        public void RemoveLast()
        {
            if (this.lastNode == null)
            {
                throw new ArgumentNullException(paramName: "No elements to remove");
            }
            --this.Count;
            this.lastNode = this.lastNode.Previous;
            if (this.lastNode != null)
            {
                this.lastNode.Next = null;
            }
        }

        public IEnumerator<T> GetEnumerator()
        {
            while (this.firstNode != null)
            {
                yield return this.firstNode.Value;
                this.firstNode = this.firstNode.Next;
            }
        }

        public T ElementAt(int index)
        {
            if (index < 0 || index >= this.Count)
            {
                throw new ArgumentOutOfRangeException(paramName: "Index was not in the range of the list");
            }

            LinkedListNode<T> node = this.firstNode;

            for (int i = 0; i < index + 1; i++)
            {
                node = node.Next;
            }

            return node.Value;
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        public bool IsEmpty()
        {
            return this.Count == 0;
        }

        private class LinkedListNode<V>
        {
            public LinkedListNode(T value)
            {
                this.Value = value;
            }

            public T Value { get; set; }

            public LinkedListNode<V> Next { get; set; }

            public LinkedListNode<V> Previous { get; set; }
        }
    }
}

File: DataStructures/js/linear-data-structures/data-structures/linked-list.js
/* globals Symbol */

class LinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.count = 0;
    }

    addLast(...values) {
        if (values.length > 1) {
            values.forEach(this.pushFront);
            return this;
        }

        const node = {
            value: values[0],
            next: null,
            prev: null,
        };

        if (this.head === null) {
            this.head = node;
            this.tail = node;
        } else {
            this.tail.next = node;
            node.prev = this.tail;
            this.tail = this.tail.next;
        }

        this.count += 1;

        return this;
    }

    addFirst(...values) {
        if (values.length > 1) {
            values.forEach(this.pushFront);
            return this;
        }
        const node = {
            value: values[0],
            next: null,
            prev: null,
        };

        if (this.head === null) {
            this.head = node;
            this.tail = node;
        } else {
            this.head.prev = node;
            node.next = this.head;
            this.head = this.head.prev;
        }

        this.count += 1;

        return this;
    }

    removeFirst() {
        if (this.head === null) {
            return null;
        }
        this.count -= 1;

        const value = this.head.value;
        this.head = this.head.next;

        if (this.head === null) {
            this.head = null;
            this.tail = null;
            return value;
        }

        this.head.prev = null;

        return value;
    }

    removeLast() {
        if (this.tail === null) {
            return null;
        }

        this.count -= 1;

        const value = this.tail.value;
        this.tail = this.tail.prev;

        if (this.tail === null) {
            this.head = null;
            this.tail = null;
            return value;
        }

        this.tail.next = null;
        return value;
    }

    isEmpty() {
        return this.count === 0;
    }

    [Symbol.iterator]() {
        let node = this.head;

        function* iterator() {
            while (node !== null) {
                yield node.value;
                node = node.next;
            }
        }

        return iterator();
    }
}

module.exports = LinkedList;